sql/pgwire_test.go

```diff
@@ -316,6 +313,103 @@ func TestPGPrepared(t *testing.T) {
 	}
 }
 
+func TestPGPreparedExec(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	var base preparedTest
+	queryTests := []struct {
+		query string
+		tests []preparedTest
+	}{
+		{
+			"CREATE DATABASE d",
+			nil,
+		},
+		{
+			"CREATE TABLE d.t (i INT, s STRING, d INT)",
+			nil,
+		},
+		{
+			"INSERT INTO d.t VALUES ($1, $2, $3)",
+			[]preparedTest{
+				base.Params(1, "one", 2),
+				base.Params("two", 2, 2).Error(`pq: param $1: strconv.ParseInt: parsing "two": invalid syntax`),
+			},
+		},
+		{
+			"UPDATE d.t SET s = $1, i = i + $2, d = 1 + $3 WHERE i = $4",
+			[]preparedTest{
+				base.Params(4, 3, 2, 1),
+			},
+		},
+		{
+			"DELETE FROM d.t WHERE s = $1 and i = $2 and d = 2 + $3",
+			[]preparedTest{
+				base.Params(1, 2, 3),
+			},
+		},
+	}
+
+	s := server.StartTestServer(t)
+	defer s.Stop()
+
+	pgUrl, cleanupFn := sqlutils.PGUrl(t, s, security.RootUser, os.TempDir(), "TestPGPrepared")
+	defer cleanupFn()
+
+	db, err := sql.Open("postgres", pgUrl.String())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer db.Close()
+
+	for _, queryTest := range queryTests {
+		if queryTest.tests == nil {
+			_, err := db.Exec(queryTest.query)
+			if err != nil {
+				t.Fatal(err)
+			}
+			continue
+		}
+		stmt, err := db.Prepare(queryTest.query)
+		if err != nil {
+			t.Fatalf("prepare error: %s: %s", queryTest.query, err)
+		}
+		for _, test := range queryTest.tests {
+			_, err := stmt.Exec(test.params...)
+			if err != nil {
+				if test.error == "" {
+					t.Fatalf("%s: %#v: unexpected error: %s", queryTest.query, test.params, err)
+				} else if test.error != err.Error() {
+					t.Fatalf("%s: %#v: expected error: %s, got %s", queryTest.query, test.params, test.error, err)
+				}
+			}
+			if test.error != "" && err == nil {
+				t.Fatalf("expected error: %s: %#v", queryTest.query, test.params)
+			}
+		}
+		if err := stmt.Close(); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	testFailures := map[string]string{
+		"CREATE TABLE $1 (id INT)":  "pq: syntax error at or near \"1\"\nCREATE TABLE $1 (id INT)\n             ^\n",
+		"DROP TABLE t":              "pq: prepare statement not supported: *parser.DropTable",
```

Comment:
Can we change this error so that it references `DROP TABLE` instead of `*parser.DropTable`? We should try to avoid leaking implementation details like the names of our internal structures into the end-user error messages.